consensus.go

```diff
@@ -0,0 +1,240 @@
+// preferences for each agenda of the latest supported stake version for the
+// selected wallet.
+func (c *Consensus) GetVoteChoices(ctx context.Context, hash string, walletID int) (*GetVoteChoicesResult, error) {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return nil, fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]VoteChoice, len(agendas)),
+	}
+
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		resp.Choices[i] = VoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agendas[i].Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for j := range agendas[i].Vote.Choices {
+			if choices[i].ChoiceID == agendas[i].Vote.Choices[j].Id {
+				resp.Choices[i].ChoiceDescription = agendas[i].Vote.Choices[j].Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+// SetVoteChoice handles a setvotechoice request by modifying the preferred
+// choice for a voting agenda.
+//
+// If a VSP host is configured in the application settings, the voting
+// preferences will also be set with the VSP.
+func (c *Consensus) SetVoteChoice(walletID int, vspPubKey []byte, vspHost, agendaID, choiceID, hash, passphrase string) error {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	err := wallet.UnlockWallet([]byte(passphrase))
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
+	if err != nil {
+		return err
+	}
+
+	if vspHost == "" {
+		return fmt.Errorf("request requires a vspHost but no vspHost was provided")
+	}
+
+	vspClient, err := wallet.LookupVSP(vspHost)
+	if err != nil && !errors.Is(err, errors.NotExist) {
+		return err
+	} else {
+		vspClient, err = wallet.VSP(vspHost, base64.StdEncoding.EncodeToString(vspPubKey))
+		if err != nil {
+			return err
+		}
+	}
+
+	// If ticket hash is provided, then set vote choice for the selected ticket
+	if ticketHash != nil {
+		err = vspClient.SetVoteChoice(ctx, ticketHash, choice)
```

Comment:
If a ticket hash is provided, then it must be registered with this vsp. Is it confirmed that the vsp host provided has control of the ticket provided? Maybe better to get the actual, correct vsp host and pubkey from the wallet using `wal.VSPTicketInfo`. In that case, the caller shouldn't provide a host or pubkey, the code should get it from the wallet using the provided ticket hash.

Reply:
![Screenshot from 2022-02-07 23-18-58](https://user-images.githubusercontent.com/25265396/152881816-4186aea7-41e2-4d8e-91c4-6506fccbf08d.png)

the `VSPTicketInfo` struct doesn't contain the pubkey

Reply:
![Screenshot from 2022-02-07 23-18-58](https://user-images.githubusercontent.com/25265396/152881816-4186aea7-41e2-4d8e-91c4-6506fccbf08d.png)

the `VSPTicketInfo` struct doesn't contain the pubkey

Reply:
I think I meant `wal.Internal().VSPTicketInfo()`. That returns
```
type VSPTicket struct {
	FeeHash     chainhash.Hash
	FeeTxStatus uint32
	VSPHostID   uint32
	Host        string
	PubKey      []byte
}
```